### PR TITLE
Benchmarks: Add benchmark - Add source code of DirectxGPUCopy microbenchmark

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Add source code of DirectxGPUCopy microbenchmark.

### CODE REVIEW

**Review by juihaochiang — APPROVED** [review]

File: superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/d3dx12.h [hunk 1]
```diff
@@ -0,0 +1,4087 @@
+//*********************************************************
```
Comment (cp5555) [inline superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/d3dx12.h:1]: Is it written by us?

File: superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+void D3D12Timer::resolveQueryToCPU(ID3D12GraphicsCommandList *pCommandList, UINT timestampPairIndex) {
+    pCommandList->ResolveQueryData(queryHeap_, D3D12_QUERY_TYPE_TIMESTAMP, timestampPairIndex * 2, 2, queryResourceCPU_,
+                                   sizeof(GPUTimestampPair) * timestampPairIndex);
+}
+
+// Get start and end timestamp pair.
+GPUTimestampPair D3D12Timer::getTimestampPair(UINT timestampPairIndex) {
+    GPUTimestampPair p{};
+
+    {
```
Comment (cp5555) [inline superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.cpp:85]: Why do we need this?

File: superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+void D3D12Timer::stop(ID3D12GraphicsCommandList *pCommandList, UINT timestampPairIndex) {
+    active_ = false;
+
+    pCommandList->EndQuery(queryHeap_, D3D12_QUERY_TYPE_TIMESTAMP, timestampPairIndex * 2 + 1);
+}
+
+// Resolve query data. Write query to device memory. Make sure to wait for query to finish before resolving data.
+void D3D12Timer::resolveQueryToCPU(ID3D12GraphicsCommandList *pCommandList, UINT timestampPairIndex) {
+    pCommandList->ResolveQueryData(queryHeap_, D3D12_QUERY_TYPE_TIMESTAMP, timestampPairIndex * 2, 2, queryResourceCPU_,
+                                   sizeof(GPUTimestampPair) * timestampPairIndex);
```
Comment (cp5555) [inline superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.cpp:78]: It's strange for formating

File: superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+#include <d3d12.h>
+
+namespace D3D12 {
+struct GPUTimestampPair {
+    UINT64 Start;
```
Comment (cp5555) [inline superbench/benchmarks/directx_benchmarks/DirectXGPUCopy/D3D12Timer.h:9]: 2-space for TAB?

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Options.h [hunk 5]
```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <iostream>
+#include <sstream>
+#include <string>
+
+struct Options {
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Options.h:10]: >Options [](http://example.com/codeflow?start=7&length=7)

there are lot of same functions as the class in another PR. How about we create a base function and create two derived for the two applications.

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Main.cpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <iostream>
+#include <sstream>
+
+#include "GPUCopyBw.h"
+
+int main(int argc, char *argv[]) {
+    Options *opts = new Options(argc, argv);
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Main.cpp:10]: > Options *opts = new Options(argc, argv); [](http://example.com/codeflow?start=3&length=41)

you are creating some objects in this file. It is OK for this application because the application quits when finishing its work, but I highly suggesting you to use smart pointer like unique_ptr to replace the raw pointers.

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.h [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  private:
+    // Pipeline objects.
+    ComPtr<ID3D12Device> m_device = nullptr;
+    ComPtr<ID3D12CommandAllocator> m_commandAllocator = nullptr;
+    ComPtr<ID3D12CommandQueue> m_commandQueue = nullptr;
+    ComPtr<ID3D12GraphicsCommandList> m_commandList = nullptr;
+
+    // App resources.
+    // Pointer of CPU size resource.
+    UINT8 *m_pDataBegin = nullptr;
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.h:130]: >UINT8 *m_pDataBegin = nullptr; [](http://example.com/codeflow?start=4&length=30)

using smart pointer?

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    ThrowIfFailed(CreateDXGIFactory2(dxgiFactoryFlags, IID_PPV_ARGS(&factory)));
+
+    ComPtr<IDXGIAdapter1> hardwareAdapter;
+    GetHardwareAdapter(factory.Get(), &hardwareAdapter);
+
+    ThrowIfFailed(D3D12CreateDevice(hardwareAdapter.Get(), D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&m_device)));
+
+    D3D12_COMMAND_QUEUE_DESC cqd3 = {};
+    cqd3.Type = D3D12_COMMAND_LIST_TYPE_COPY;
+    m_device->CreateCommandQueue(&cqd3, IID_PPV_ARGS(&m_commandQueue));
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp:233]: >CreateCommandQueue [](http://example.com/codeflow?start=14&length=18)

actually functions like this returns HRESULT to show if it executes correctly. I guess you may check the return value to make sure it executes without any errors before go ahead.

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Options.h (hunk 5 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Options.h]: ok, sure, updated to move the base class options.h to directx_utils, please check it

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Main.cpp (hunk 6 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/Main.cpp]: updated, please help check it I use them in a correct way

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.h (hunk 7 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.h]: updated

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    ID3D12CommandList *listsToExecute[] = {m_commandList.Get()};
+    this->m_commandQueue->ExecuteCommandLists(ARRAYSIZE(listsToExecute), listsToExecute);
+    // Signal and increment the fence value.
+    const UINT64 fenceL = m_copyFenceValue;
+    m_commandQueue->Signal(m_copyFence.Get(), fenceL);
+    m_copyFenceValue++;
+    // Wait until command queue is done.
+    if (m_copyFence->GetCompletedValue() < fenceL) {
+        m_copyFence->SetEventOnCompletion(fenceL, m_copyEventHandle);
+        WaitForSingleObject(m_copyEventHandle, INFINITE);
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp]: oh, updated similar with previous PR

File: superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp (hunk 8 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp]: thanks, added throwiffailed to all functions which returns HRESULT

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h [hunk 10]
```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/directx_utils/Options.h:1]: I think it should be conflicting with another PR?

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+     * @brief Get the int type value of cmd line argument.
+     * @param option the cmd line argument.
+     * @param defaults the default value.
+     * @return int the int type value of cmd line argument 'option'.
+     */
+    int get_cmd_line_argument_int(const std::string &option, int defaults) {
+        if (char *value = get_cmd_option(option)) {
+            try {
+                return std::stoi(value);
+            } catch (const std::exception &e) {
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_utils/Options.h:38]: well, if there is exception, you are going to return the default value, right? If it is so, I guess you can return defaults directly in the catch clause instead of using exit(1), but if you need to quit the whole process caused by that, you can keep it 😊

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+     * @return unsigned long long the unsigned long long type value of cmd line argument 'option'.
+     */
+    unsigned long long get_cmd_line_argument_ulonglong(const std::string &option, unsigned long long defaults) {
+        if (char *value = get_cmd_option(option)) {
+            try {
+                return std::stoull(value);
+            } catch (const std::exception &e) {
+                std::cout << "Error: Invalid argument - " << option << " should be unsigned long long" << e.what()
+                          << '\n';
+                exit(1);
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_utils/Options.h:59]: same here

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * @brief Get the option usage.
+     */
+    virtual void get_option_usage(){};
+
+    /**
+     * @brief Parse the arguments.
+     */
+    virtual void parse_arguments(){};
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/directx_utils/Options.h:104]: how about set them as pure virtual functions like virtual void func_name = 0, so we can not use this class directly?

**Comment by hippogr** [comment]
still some minor coments. 😊

**Review by hippogr — APPROVED** [review]
I will approve first and please go ahead after you fixes all the comments.

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h (hunk 12 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_utils/Options.h]: I think just throw it to let user know

File: superbench/benchmarks/micro_benchmarks/directx_utils/Options.h (hunk 13 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/directx_utils/Options.h]: sure, updated
